File: cpp/pl/brpc/echo_client.cpp
// Copyright (c) 2026 The Authors. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Authors: liubang (it.liubang@gmail.com)
// Created: 2026/05/06 21:10

#include <brpc/channel.h>
#include <gflags/gflags.h>

#include "cpp/pl/brpc/proto/echo.pb.h"

DEFINE_string(server, "0.0.0.0:8000", "IP Address of server");
DEFINE_string(load_balancer, "", "The algorithm for load balancing");
DEFINE_int32(timeout_ms, 100, "RPC timeout in milliseconds");
DEFINE_int32(max_retry, 3, "Max retries(not including the first RPC)");
DEFINE_int32(request_count, 10, "Number of requests to send");

int main(int argc, char* argv[]) {
    gflags::ParseCommandLineFlags(&argc, &argv, true);

    ::brpc::Channel channel;

    ::brpc::ChannelOptions options;
    options.timeout_ms = FLAGS_timeout_ms;
    options.max_retry = FLAGS_max_retry;

    // 初始化 channel，连接到 server
    if (channel.Init(FLAGS_server.c_str(), FLAGS_load_balancer.c_str(), &options) != 0) {
        LOG(ERROR) << "Fail to initialize channel";
        return -1;
    }

    // 通过 channel 访问 EchoService，不需要手动创建连接
    pl::brpc::proto::EchoService_Stub stub(&channel);

    for (int i = 0; i < FLAGS_request_count; ++i) {
        pl::brpc::proto::EchoRequest request;
        pl::brpc::proto::EchoResponse response;
        ::brpc::Controller cntl;

        request.set_message("hello brpc " + std::to_string(i));

        // 同步调用，cntl 中携带了请求的上下文和结果
        stub.Echo(&cntl, &request, &response, nullptr);

        if (!cntl.Failed()) {
            LOG(INFO) << "Received response from " << cntl.remote_side() << " to "
                      << cntl.local_side() << ": " << response.message()
                      << " latency=" << cntl.latency_us() << "us";
        } else {
            LOG(WARNING) << cntl.ErrorText();
        }
    }

    return 0;
}

File: java/pl/brpc/echo/src/main/java/pl/brpc/echo/EchoClient.java
// Copyright (c) 2026 The Authors. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Authors: liubang (it.liubang@gmail.com)
// Created: 2026/06/21

package pl.brpc.echo;

import com.baidu.cloud.starlight.api.rpc.StarlightClient;
import com.baidu.cloud.starlight.api.rpc.config.ServiceConfig;
import com.baidu.cloud.starlight.api.rpc.config.TransportConfig;
import com.baidu.cloud.starlight.core.rpc.SingleStarlightClient;
import com.baidu.cloud.starlight.core.rpc.proxy.JDKProxyFactory;

import pl.grpc.proto.EchoRequest;
import pl.grpc.proto.EchoResponse;
import pl.grpc.proto.HealthRequest;
import pl.grpc.proto.HealthResponse;

/**
 * Starlight echo client — calls the 2 unary RPCs (Echo + HealthCheck)
 * over the brpc binary protocol.
 * <p>
 * Can talk to either a Java Starlight server or a C++ brpc server.
 */
public class EchoClient {

    public static void main(String[] args) {
        String host = args.length > 0 ? args[0] : "localhost";
        int port = args.length > 1 ? Integer.parseInt(args[1]) : 8005;

        TransportConfig transportConfig = new TransportConfig();
        StarlightClient client = new SingleStarlightClient(host, port, transportConfig);
        client.init();

        ServiceConfig serviceConfig = new ServiceConfig();
        serviceConfig.setProtocol("brpc");
        // For cross-language interop with C++ brpc, set the proto service name.
        // When talking to a Java Starlight server, javaInterface.getName() is used by default.
        // brpc service FQN: pl.brpc.echo.EchoService
        serviceConfig.setServiceId("pl.brpc.echo.EchoService");
        serviceConfig.setSerializeMode("pb2-std"); // standard protobuf for C++ compat

        JDKProxyFactory proxyFactory = new JDKProxyFactory();
        EchoService echoService = proxyFactory.getProxy(
                EchoService.class, serviceConfig, client);

        try {
            // --- HealthCheck ---
            System.out.println("============ HealthCheck ============");
            HealthRequest healthRequest = HealthRequest.newBuilder().build();
            HealthResponse healthResponse = echoService.HealthCheck(healthRequest);
            System.out.println("[HealthCheck] status="
                    + healthResponse.getStatus()
                    + " | server=" + healthResponse.getServerId()
                    + " | version=" + healthResponse.getVersion()
                    + " | uptime=" + healthResponse.getUptimeSeconds() + "s");

            // --- Unary Echo ---
            System.out.println("============ Unary Echo =============");
            long sentUs = System.currentTimeMillis() * 1000;
            EchoRequest echoRequest = EchoRequest.newBuilder()
                    .setMessage("Hello from Java Starlight client!")
                    .setTimestampUs(sentUs)
                    .putHeaders("client", "java-starlight")
                    .build();
            EchoResponse echoResponse = echoService.Echo(echoRequest);
            long rttUs = System.currentTimeMillis() * 1000 - sentUs;
            System.out.println("[Echo] response: " + echoResponse.getMessage()
                    + " | rtt_us=" + rttUs
                    + " | server=" + echoResponse.getServerId());

        } catch (Exception e) {
            System.err.println("[EchoClient] RPC failed: " + e.getMessage());
            e.printStackTrace();
        } finally {
            client.destroy();
        }
    }
}
